backend/app/matrix/external_url.py

```diff
@@ -0,0 +1,70 @@
+import random
+import string
+
+from app.core.background_runner import matrix_bot_runner
+
+
+class ExternalUrl:
+    """
+    ExternalUrl class which abstracts out the interactions between the Bot Client and the API routes.
+    """
+
+    def __init__(self):
+        self._characters = string.ascii_letters + string.digits
+        self._event_type = "matrix-cerberus.external_url"
+
+    def _generate_url_code(self, N: int = 8):
+        return "".join(random.choices(self._characters, k=N))
+
+    async def parse_data_and_save(self, matrix_homeserver, external_url_data):
+        data = {"content": external_url_data}
+
+        await matrix_bot_runner.client.put_account_data(self._event_type, matrix_homeserver, data)
+
+    async def generate_url(self, matrix_homeserver: str, room_id: str, use_once_only: bool):
+        data = await matrix_bot_runner.client.get_account_data(self._event_type, matrix_homeserver)
+        external_url_data = data.content
+        url_code = self._generate_url_code()
+        while url_code in external_url_data:
+            url_code = self._generate_url_code()
+
+        external_url_data[url_code] = {"room_id": room_id, "use_once_only": use_once_only}
+
+        await self.parse_data_and_save(matrix_homeserver, external_url_data)
+        return url_code
+
+    async def get_room_invite(self, matrix_homeserver: str, url_code: str, user_id: str):
+        data = await matrix_bot_runner.client.get_account_data(self._event_type, matrix_homeserver)
```

Comment:
We are going to want to think about a cache for all these get account data calls sooner rather than later. Account data changes come down in `/sync` so we could consider running a sync loop to blow out our cache. That's a more advanced solution, for the moment we could setup a cache where we assume no other device is modifying the bot users account data.